File: dynamic_parameter_py/dynamic_parameter_py/lenient_multiplier.py
# Copyright 2026 Jasper van Brakel
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from collections.abc import Sequence

from example_interfaces.msg import Float64
from rcl_interfaces.msg import ParameterDescriptor
from rcl_interfaces.msg import SetParametersResult

import rclpy
from rclpy.executors import ExternalShutdownException
from rclpy.node import Node
from rclpy.parameter import Parameter

NUMERIC_TYPES = {Parameter.Type.DOUBLE, Parameter.Type.INTEGER}


class LenientMultiplier(Node):
    """A multiplier node with a double parameter which also excepts ints."""

    def __init__(self) -> None:
        super().__init__('lenient_multiplier')

        # Add a parameter to control if the demo is lenient or strict
        is_lenient: bool = self.declare_parameter(
            'is_lenient',
            True,
            descriptor=ParameterDescriptor(
                description="Toggle if the 'multiplier' parameter is lenient",
                read_only=True,
            ),
        ).value

        if is_lenient:
            self.get_logger().info("The 'multiplier' parameter is lenient")
            self.add_on_set_parameters_callback(self._validate_parameters_callback)
        else:
            self.get_logger().warning("The 'multiplier' parameter is statically typed!")

        self.add_post_set_parameters_callback(self._post_set_parameters_callback)

        self.declare_parameter('multiplier', value=1.0, descriptor=ParameterDescriptor(
            dynamic_typing=is_lenient,
            additional_constraints=('Must be a valid numeric type (DOUBLE or INTEGER)'
                                    if is_lenient else None)
        ))

        self._multiplied_pub = self.create_publisher(Float64, 'multiplied_number', 10)
        self._number_sub = self.create_subscription(Float64, 'number', self._number_callback, 10)

    def _validate_parameters_callback(self, params: Sequence[Parameter]) -> SetParametersResult:
        for param in params:
            # Reject updates to the multiplier parameter if the type is not numeric
            if param.name == 'multiplier' and param.type_ not in NUMERIC_TYPES:
                return SetParametersResult(
                    successful=False,
                    reason=(
                        "'multiplier' must be a numeric type (DOUBLE or INTEGER), "
                        f"but was of type '{param.type_.name}'"
                    )
                )

        return SetParametersResult(successful=True)

    def _post_set_parameters_callback(self, params: Sequence[Parameter]) -> None:
        for param in params:
            if param.name == 'multiplier':
                self.get_logger().info(f'The multiplier has been set to {param.value!r}')

    def _number_callback(self, number: Float64) -> None:
        # No extra conversion is necessary as Python implicitly casts numeric
        # types when necessary in multiplication.
        multiplier = self.get_parameter('multiplier').value

        self.get_logger().info(f'Multiplying {number.data!r} with multiplier {multiplier!r}')

        msg = Float64(data=number.data*multiplier)
        self._multiplied_pub.publish(msg)


def main(args=None) -> None:
    try:
        # NOTE: init context manager available starting from Kilted
        rclpy.init(args=args)
        node = LenientMultiplier()

        rclpy.spin(node)

    except (KeyboardInterrupt, ExternalShutdownException):
        pass
    finally:
        rclpy.try_shutdown()


if __name__ == '__main__':
    main()

File: dynamic_parameter_cpp/src/lenient_multiplier.cpp
// Copyright 2026 Jasper van Brakel
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <functional>
#include <sstream>
#include <vector>

#include "dynamic_parameter_cpp/lenient_multiplier.hpp"

#include "rclcpp/logging.hpp"
#include "rclcpp_components/register_node_macro.hpp"

#include "example_interfaces/msg/float64.hpp"
#include "rcl_interfaces/msg/parameter_descriptor.hpp"
#include "rcl_interfaces/msg/set_parameters_result.hpp"

using namespace std::placeholders;
using namespace std::chrono_literals;

namespace dynamic_parameter_cpp
{

LenientMultiplier::LenientMultiplier(const rclcpp::NodeOptions & options)
: Node("lenient_multiplier", options)
{
  using rcl_interfaces::msg::ParameterDescriptor;

  // A parameter to control if the demo node is lenient or strict
  auto is_lenient_descriptor = ParameterDescriptor();
  is_lenient_descriptor.description =
    "Toggle if the 'multiplier' parameter is lenient";
  is_lenient_descriptor.read_only = true;
  bool is_lenient =
    declare_parameter("is_lenient", true, is_lenient_descriptor);

  if (is_lenient) {
    RCLCPP_INFO(get_logger(), "The 'multiplier' parameter is lenient");
    on_set_parameter_callback_handle_ = add_on_set_parameters_callback(
        std::bind(&LenientMultiplier::validate_parameters_callback, this, _1));
  } else {
    RCLCPP_WARN(get_logger(),
                "The 'multiplier' parameter is statically typed!");
  }

  // Add a post set callback so changes can bne logged easily
  post_set_parameter_callback_handle_ =
    add_post_set_parameters_callback(std::bind(&LenientMultiplier::post_set_parameters_callback,
      this, _1));

  auto multiplier_descriptor = ParameterDescriptor();
  multiplier_descriptor.dynamic_typing = is_lenient;
  if (is_lenient) {
    multiplier_descriptor.additional_constraints =
      "Must be a valid numeric type ('double' or 'integer')";
  }
  declare_parameter("multiplier", 1.0, multiplier_descriptor);

  multiplied_pub_ = create_publisher<example_interfaces::msg::Float64>("multiplied_number", 10);
  number_sub_ = create_subscription<example_interfaces::msg::Float64>("number", 10,
      std::bind(&LenientMultiplier::number_callback, this, _1));
}

rcl_interfaces::msg::SetParametersResult
LenientMultiplier::validate_parameters_callback(
  const std::vector<rclcpp::Parameter> & parameters) const
{
  auto result = rcl_interfaces::msg::SetParametersResult();

  for (const auto & param : parameters) {
    if (const auto name = param.get_name(); name == "multiplier") {
      if (const auto param_type = param.get_type();
        param_type != rclcpp::PARAMETER_DOUBLE &&
        param_type != rclcpp::PARAMETER_INTEGER)
      {
        std::ostringstream reason;
        reason << "'" << name
               << "' must be a numeric type ('double' or 'integer'), but was of type '"
               << param_type << "'";

        result.successful = false;
        result.reason = reason.str();

        return result;
      }
    }
  }

  result.successful = true;
  return result;
}

void LenientMultiplier::post_set_parameters_callback(
  const std::vector<rclcpp::Parameter> & parameters) const
{
  for (const auto & param : parameters) {
    if (const auto name = param.get_name(); name == "multiplier") {
      RCLCPP_INFO_STREAM(get_logger(),
          "The '" << name << "' parameter has been set to '" <<
          rclcpp::to_string(param.get_parameter_value()) << "'");
    }
  }
}

void LenientMultiplier::number_callback(
  const example_interfaces::msg::Float64::ConstSharedPtr number) const
{
  auto multiplier_param = get_parameter("multiplier");

  RCLCPP_INFO_STREAM(get_logger(),
      "Multiplying " << number->data << " with multiplier " <<
      rclcpp::to_string(multiplier_param.get_parameter_value()));

  double multiplier;
  if (multiplier_param.get_type() == rclcpp::PARAMETER_DOUBLE) {
    multiplier = multiplier_param.as_double();
  } else {
    multiplier = static_cast<double>(multiplier_param.as_int());
  }

  auto msg = example_interfaces::msg::Float64();
  msg.data = multiplier * number->data;
  multiplied_pub_->publish(msg);
}

}  // namespace dynamic_parameter_cpp

RCLCPP_COMPONENTS_REGISTER_NODE(dynamic_parameter_cpp::LenientMultiplier)
